### PR TITLE
feat: add experimental transition focus manager

### LINKED ISSUE #23650: bug: ios, voiceover does not move to correct location after page transition
### Prequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already report this problem, without success.

### Ionic Framework Version

- [ ] v4.x
- [X] v5.x
- [X] v6.x

### Current Behavior

Hi!

I stumbled upon a bug while implementing some accessibility functionality in an application

When navigating from page A to page B with voice-over within my application, the following issue appears. Upon entry to page B, the voice over-gets stuck on page A. Voice-over does not detect the correct contents inside page B. It looks like the voice-over keeps focusing on the recently clicked element from the previous page (A).

I have recorded functionality behaviour in the video linked below. 

https://user-images.githubusercontent.com/54717333/125956650-0ad6d929-7623-43a1-b4f0-54ad0d91b68e.mov

### Expected Behavior

When redirecting from "test page1" (A) to "test page2" (B) voice-over should immediately detect the first contents within the pages body so that the voice-over starts reading from the top of the page.

### Steps to Reproduce

**Project initiation**
* Download the repo
* Open your CLI and navigate to the project folder
* Type "npm install" and "ionic serve"

**(macOS)**
* Go to system preferences > accessibility > shortcut > make sure every box is checked
* Open the browser window that contains the project server > right click > inspect > toggle device toolbar "iphone X"
* Press CMD while quickly pressing the "touch id" button 3 times
* Refresh the browser window that contains the project server

**Voice-over navigation**
* CAPS-lock + shift + down-arrow to enter the current area
* CAPS-lock + shift + up-arrow to exit the current area
* CAPS-lock + shift + right-arrow to skip the current area
* CAPS-lock + space to "click" on a clickable link/button/item

### Code Reproduction URL

https://github.com/Waligustav/ionicbugtest

### Ionic Info

[WARN] Error loading @capacitor/ios package.json: Error: Cannot find module
       '@capacitor/ios/package'
       
       Require stack:
       -
       /Users/walibjork/.nvm/versions/node/v14.7.0/lib/node_modules/@ionic/cli/lib/project/index.js
       -
       /Users/walibjork/.nvm/versions/node/v14.7.0/lib/node_modules/@ionic/cli/lib/index.js
       - /Users/walibjork/.nvm/versions/node/v14.7.0/lib/node_modules/@ionic/cli/index.js
       - /Users/walibjork/.nvm/versions/node/v14.7.0/lib/node_modules/@ionic/cli/bin/ionic
[WARN] Error loading @capacitor/android package.json: Error: Cannot find module
       '@capacitor/android/package'
       
       Require stack:
       -
       /Users/walibjork/.nvm/versions/node/v14.7.0/lib/node_modules/@ionic/cli/lib/project/index.js
       -
       /Users/walibjork/.nvm/versions/node/v14.7.0/lib/node_modules/@ionic/cli/lib/index.js
       - /Users/walibjork/.nvm/versions/node/v14.7.0/lib/node_modules/@ionic/cli/index.js
       - /Users/walibjork/.nvm/versions/node/v14.7.0/lib/node_modules/@ionic/cli/bin/ionic

Ionic:

   Ionic CLI       : 6.16.3 (/Users/walibjork/.nvm/versions/node/v14.7.0/lib/node_modules/@ionic/cli)
   Ionic Framework : @ionic/vue 5.6.11

Capacitor:

   Capacitor CLI      : 3.1.1
   @capacitor/android : not installed
   @capacitor/core    : 3.1.1
   @capacitor/ios     : not installed

Utility:

   cordova-res : not installed globally
   native-run  : 1.4.0

System:

   NodeJS : v14.7.0 (/Users/walibjork/.nvm/versions/node/v14.7.0/bin/node)
   npm    : 6.14.7
   OS     : macOS Catalina

### Additional Information

I upgraded a copy of the project to the Ionic 6 beta and tested the same functionality without any positive results.

I also tried this on a physical Iphone and voice-over seems to behave similarly.
Let me know if you want me to post more information about testing this issue on a physical device and i will provide you with more information.

### PR DESCRIPTION
**Pull request by liamdebeasi**
Issue number: resolves #23650

---------

## What is the current behavior?

In traditional native applications, navigation will inform screen readers that the view has changed. This allows screen readers to focus the correct view. In a single page app on the web, this same concept does not exist. As a result, transitioning from Page A to Page B results in screen reader focus remaining on Page A. This means that users who rely on screen readers are not informed of view changes.

Currently, developers are responsible for implementing this on their own.

## What is the new behavior?

- Introduces a new focus manager priority global config. When defined, the app developer can specify which area of the view focus should be moved to when the transition ends. The developer does this by specifying areas in order of priority which allows for fallbacks in the event that a particular UI component (such as a header) does not exist on a view.

There is some risk here by managing focus for the application. As a result, this feature is considered experimental and disabled by default. The team should collect feedback based on usage and enable it by default when they feel this feature is stable enough.

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

⚠️ Due to the `tsconfig.json` change, reviewers should restart the Stencil dev server when checking out these changes locally.

Reviewers: Please test both of the test template files on physical iOS and Android device with VoiceOver and TalkBack enabled, respectively.

Docs Link: https://github.com/ionic-team/ionic-docs/pull/3627

### CODE REVIEW

File: core/src/utils/transition/index.ts [hunk 1]
```diff
@@ -7,11 +7,14 @@ import {
   LIFECYCLE_WILL_LEAVE,
 } from '../../components/nav/constants';
 import type { NavOptions, NavDirection } from '../../components/nav/nav-interface';
+import { config } from '../../global/config';
```
Comment (sean-perkins) [inline core/src/utils/transition/index.ts:10]: Should this use the `@global/config` import alias as well?

Comment (liamdebeasi) [reply core/src/utils/transition/index.ts]: Fixed in https://github.com/ionic-team/ionic-framework/pull/29400/commits/2c84f632c6d5848aa6949e48660b31e346e81fa7

File: core/src/utils/focus-controller/test/ionic/focus-controller.e2e.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+configs({ modes: ['md'], directions: ['ltr'] }).forEach(({ title, config }) => {
+  test.describe(title('focus controller: ionic components'), () => {
+    test.beforeEach(async ({ page }) => {
+      await page.goto('/src/utils/focus-controller/test/ionic', config);
+    });
+    test('should focus heading', async ({ page }) => {
+      const goToPageOneButton = page.locator('page-root ion-button.page-one');
+      const nav = page.locator('ion-nav');
+      const ionNavDidChange = await (nav as any).spyOnEvent('ionNavDidChange');
```
Comment (liamdebeasi) [inline core/src/utils/focus-controller/test/ionic/focus-controller.e2e.ts:12]: I need to figure out why `spyOnEvent` does not exist on the type interface for our custom Locator, but the method does indeed exist.

Comment (sean-perkins) [reply core/src/utils/focus-controller/test/ionic/focus-controller.e2e.ts:12]: You need to cast:
```ts
import type { E2ELocator } from '@utils/test/playwright';
...
const nav = page.locator('ion-nav') as E2ELocator;
```

File: core/src/css/core.scss [hunk 3]
```diff
@@ -395,6 +395,23 @@ ion-input input::-webkit-date-and-time-value {
...
+ * by the focus manager. We are intentionally selective about which elements this
+ * applies to so we do not accidentally override outlines set by the developer.
+ */
+[ion-last-focus],
+header[tabindex="-1"]:focus,
+[role="banner"][tabindex="-1"]:focus,
+main[tabindex="-1"]:focus,
+[role="main"][tabindex="-1"]:focus,
+h1[tabindex="-1"]:focus,
+[role="heading"][aria-level="1"][tabindex="-1"]:focus {
```
Comment (thetaPC) [inline core/src/css/core.scss:410]: Could we do the following to lessen the amount of lines? Or is it more important to point out which elements are being targeted?

```suggestion
[ion-last-focus],
*[tabindex="-1"]:focus {
```

Comment (liamdebeasi) [reply core/src/css/core.scss:410]: This changes which elements are going to be selected. The focus manager works on specific semantic elements/landmarks, but the proposed change would apply `outline: none` to any focused element that has `tabindex="-1"`. The particular elements that the focus manager targets is outlined on the docs: https://ionic-docs-git-focus-manager-docs-ionic1.vercel.app/docs/developing/managing-focus#content-types

**Review by thetaPC — APPROVED** [review]
LGTM, worked fine on iOS and Android physical devices.

**Pull request merged**